File: java-workshop/part-1-powermock/src/test/java/service/DiscountCodeServiceTest.java
package service;

import domain.DiscountCode;
import infrastructure.DatabaseDiscountCodeRepository;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mockConstruction;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class DiscountCodeServiceTest {

    @Test
    void p1_checkDiscountCode_returnsTrue_whenCodeHasAlreadyBeenUsed() {
        try (var mockedRepo = mockConstruction(DatabaseDiscountCodeRepository.class,
                (mock, ctx) -> when(mock.checkDiscountCode("CUST-001", DiscountCode.SUMMER20)).thenReturn(true))) {

            var discountCodeService = new DiscountCodeService();

            assertTrue(discountCodeService.checkDiscountCode("CUST-001", DiscountCode.SUMMER20));
        }
    }

    @Test
    void p1_checkDiscountCode_returnsFalse_whenCodeHasNotBeenUsed() {
        try (var mockedRepo = mockConstruction(DatabaseDiscountCodeRepository.class,
                (mock, ctx) -> when(mock.checkDiscountCode("CUST-001", DiscountCode.SUMMER20)).thenReturn(false))) {

            var discountCodeService = new DiscountCodeService();

            assertFalse(discountCodeService.checkDiscountCode("CUST-001", DiscountCode.SUMMER20));
        }
    }

    @Test
    void p1_markAsUsed_delegatesToRepository() {
        try (var mockedRepo = mockConstruction(DatabaseDiscountCodeRepository.class)) {

            var discountCodeService = new DiscountCodeService();
            discountCodeService.markAsUsed("CUST-001", DiscountCode.SUMMER20);

            verify(mockedRepo.constructed().getFirst()).markAsUsed("CUST-001", DiscountCode.SUMMER20);
        }
    }
}

File: csharp-workshop/Part1.StaticDeps.Tests/Service/DiscountCodeServiceTest.cs
using Part1.StaticDeps.Domain;
using Part1.StaticDeps.Service;
using Xunit;

namespace Part1.StaticDeps.Tests.Service;

/// <summary>
/// All tests require constructor interception of
/// <see cref="Part1.StaticDeps.Infrastructure.DatabaseDiscountCodeRepository"/>.
/// In Java, Mockito.mockConstruction() intercepts every <c>new DatabaseDiscountCodeRepository()</c>
/// call inside the service constructor. In C#, NSubstitute cannot intercept constructors.
/// Use JustMock or Typemock for the equivalent capability.
/// </summary>
public class DiscountCodeServiceTest
{
    [Fact(Skip = "mockConstruction requires JustMock/Typemock in C#. Java equivalent uses Mockito.mockConstruction().")]
    public void P1_CheckDiscountCode_ReturnsTrue_WhenCodeHasAlreadyBeenUsed()
    {
        // Given
        // mockConstruction(DatabaseDiscountCodeRepository) stubs
        //   CheckDiscountCode("CUST-001", DiscountCode.SUMMER20) → true

        // When
        // discountCodeService.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20)

        // Then result == true
    }

    [Fact(Skip = "mockConstruction requires JustMock/Typemock in C#. Java equivalent uses Mockito.mockConstruction().")]
    public void P1_CheckDiscountCode_ReturnsFalse_WhenCodeHasNotBeenUsed()
    {
        // Given
        // mockConstruction(DatabaseDiscountCodeRepository) stubs
        //   CheckDiscountCode("CUST-001", DiscountCode.SUMMER20) → false

        // Then result == false
    }

    [Fact(Skip = "mockConstruction requires JustMock/Typemock in C#. Java equivalent uses Mockito.mockConstruction().")]
    public void P1_MarkAsUsed_DelegatesToRepository()
    {
        // Given mockConstruction(DatabaseDiscountCodeRepository) with no stubs

        // When discountCodeService.MarkAsUsed("CUST-001", DiscountCode.SUMMER20)

        // Then
        // constructedRepo.Received(1).MarkAsUsed("CUST-001", DiscountCode.SUMMER20)
    }
}
